### PR TITLE
OpenAssistant instruction data

### PR DESCRIPTION
**Pull request by TearGosling**
Rather basic implementation of the OpenAssistant dataset. Let me know if you've got comments.

### CODE REVIEW

File: toolbox/datasets/oasst.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import json
+import os
+import typing as t
+
+from dataclasses import dataclass
+
+from toolbox.core.dataset import BaseDataset, get_path_for
+
+# Thanks GPT-4
+def get_list_of_dicts(d):
```
Comment (0x000011b) [inline toolbox/datasets/oasst.py]: This function is for exclusive use internal to this file, so I'd recommend renaming it to `_get_list_of_dicts` and moving it to the bottom of the file for consistency with the rest of the repo.

File: toolbox/datasets/oasst.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    language: str
+    tree_id: str
+
+class OpenAssistantDataset(BaseDataset[OpenAssistantDataEntry]):
+    '''
+    The OpenAssistant dataset from the OpenAssistant organization - specifically, the cleaned version.
+    https://huggingface.co/datasets/OpenAssistant/oasst1
+
+    Params:
+    kept_languages: The list of languages in OpenAssist to keep in the final dataset. Set to None to keep all languages.
```
Comment (0x000011b) [inline toolbox/datasets/oasst.py]: Bad copy-paste? I see that this param is a part of the `Task`, not the `Dataset`.

File: toolbox/datasets/oasst.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        file_path = os.path.join(root_data_path, "2023-04-12_oasst_ready.trees.jsonl")
+        
+        with open(file_path, "r", encoding="utf-8") as f:
+            for line in f:
+                entry = json.loads(line)
+                # Create a list of conversations from the different conversation paths
+                for conversation in get_list_of_dicts(entry["prompt"]):
+                    yield OpenAssistantDataEntry(
+                        conversation=conversation,
+                        language=conversation[0]["lang"],
```
Comment (0x000011b) [inline toolbox/datasets/oasst.py]: I have this exact same file on my end (`2023-04-12_oasst_ready.trees.jsonl`), but this line crashes with:

```
    language=conversation[0]["lang"],
KeyError: 'lang'
```

On my machine. Did I miss something?

File: toolbox/tasks/oasst_instruction_following.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    def __init__(self, kept_languages: t.Optional[list[str]] = None) -> None:
+        self.kept_languages = kept_languages
+        super().__init__()
+
+    def __iter__(self) -> t.Generator[Episode, None, None]:
+        # Cache parent tree id for identifier string
+        tree_id = ""
+        for conversation in OpenAssistantDataset():
+            # If the language of the conversation is not in the kept languages, skip it
+            if self.kept_languages is not None and conversation[0]['lang'] not in self.kept_languages:
```
Comment (0x000011b) [inline toolbox/tasks/oasst_instruction_following.py]: We can check `conversation.language` instead of `conversation[0]['lang']` here.

**Review by 0x000011b — CHANGES_REQUESTED** [review]
I failed to test this one on my end, but already caught some stuff by looking at the code.

**Comment by TearGosling** [comment]
Pushed some changes which hopefully resolve all these issues.

**Comment by TearGosling** [comment]
Ancient PR, closing for tidying up

**Pull request closed without merging**